ownCloud/Settings/Passcode/UnlockPasscodeManager.swift

```diff
@@ -0,0 +1,118 @@
+    }()
+
+    public override init() {
+        self.userDefaults = UserDefaults.standard
+
+        super.init()
+    }
+
+    // MARK: - Unlock device
+
+    func showPasscodeIfNeeded(viewController: UIViewController, hiddenOverlay:Bool) {
+
+        if isPasscodeActivated() {
+
+            storeDateHomeButtonPressed()
+
+            if self.passcodeViewController == nil {
+                self.passcodeViewController = PasscodeViewController(mode: PasscodeInterfaceMode.unlockPasscode, hiddenOverlay:false, handler: {
+                    DispatchQueue.main.async {
+                        self.userDefaults?.removeObject(forKey: DateHomeButtonPressedKey)
+                        viewController.dismiss(animated: true, completion: nil)
+                        self.passcodeViewController = nil
+                    }
+                })
+
+                viewController.present(self.passcodeViewController!, animated: false, completion: nil)
+            } else {
+                self.passcodeViewController?.showOverlay()
+            }
+        }
+    }
+
+    // MARK: - Interface updates
+
+    private func hideOverlay() {
+        if self.passcodeViewController != nil {
+            self.passcodeViewController?.hideOverlay()
+        }
+    }
+
+    func dismissAskedPasscodeIfDateToAskIsLower() {
+
+        if !isNeccesaryShowPasscode() {
+            if self.passcodeViewController != nil {
+                self.passcodeViewController?.dismiss(animated: true, completion: nil)
+                self.passcodeViewController = nil
+                self.userDefaults?.removeObject(forKey: DateHomeButtonPressedKey)
+            }
+        } else {
+            hideOverlay()
+        }
+    }
+
+    // MARK: - Utils
+
+    private func isPasscodeActivated() -> Bool {
+
+        var output: Bool = true
+
+        if OCAppIdentity.shared().keychain.readDataFromKeychainItem(forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) == nil {
+
+            output = false
+        }
+
+        return output
+    }
+
+    private func isNeccesaryShowPasscode() -> Bool {
+
+        var output: Bool = true
+
+        if isPasscodeActivated() {
+            if let dateData = self.userDefaults?.data(forKey: DateHomeButtonPressedKey) {
+                if let date = NSKeyedUnarchiver.unarchiveObject(with: dateData) as? Date {
+
+                    let elapsedSeconds = Date().timeIntervalSince(date)
+                    let minSecondsToAsk = self.userDefaults?.integer(forKey: SecuritySettingsFrequencyKey)
+
+                    if Int(elapsedSeconds) < minSecondsToAsk! {
+                        output = false
+                    }
+                }
+            }
+        } else {
+            output = false
+        }
+
+        return output
+    }
+    
+    func storeDateHomeButtonPressed() {
+        if OCAppIdentity.shared().keychain.readDataFromKeychainItem(forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) != nil,
```

Comment:
This is duplicate code. `isPasscodeActivated` could/should be called instead.

Reply:
You are right, removed 👍